test/controllers/tasks_controller_test.rb

```diff
@@ -106,21 +114,86 @@
     #        thing to test.
     it "can update an existing task" do
       # Your code here
+      # Instead of Task.first, do Task.create
+      existing_task = Task.create(name: "Watch TV", description: "Description for tv")
+         
+      update_task_from_data = {
+        task:{
+          name: "Watch TV",
+          description: "Descripton for tv"
+          # completed: ""
+        }
+      }
+      # Act
+     patch task_path(existing_task.id, params: update_task_from_data)
+      #     must_redirect_to root_path
+      # must_redirect_to root_path
+
+      # Assert
+      expect( Task.find_by(id: existing_task.id).name ).must_equal "Watch TV"
     end
 
     it "will redirect to the root page if given an invalid id" do
       # Your code here
+      # Arrange
+    
+      invalid_id = "-1"
+      update_task_from_data = {
+        task:{
+          name: "Watch TV",
+          description: "Descripton for tv"
+          # completed: ""
+        }
+      }
+      get task_path( invalid_id, params: update_task_from_data )
+      # Assert
+      must_respond_with :redirect
+      # must_redirect_to task_path
     end
   end
 
+
   # Complete these tests for Wave 4
   describe "destroy" do
     # Your tests go here
+      it "Deleting is successful!" do
+        # Create a Task, then find the newly created Task's id
+      
+        new_task = Task.create(name: "Watch TV", description: "Description for tv")
+        id = new_task.id
 
+        expect {
+          delete task_delete_path(id)
+        }.must_change "Task.count", 1
+
+        must_respond_with :redirect
+        must_redirect_to root_path
+
+        expect( Task.find_by(id: id) ).must_be_nil
+    end
   end
 
-  # Complete for Wave 4
-  describe "toggle_complete" do
-    # Your tests go here
+ # Complete for Wave 4
+ describe "toggle_complete" do
+    it "can mark tasks as completed " do
+      # Your code here
+      existing_task = Task.create(name: "Watch TV", description: "Description for tv")
+          
+      update_task_from_data = {
+        task:{
+          name: "Watch TV",
+          description: "Descripton for tv",
+          completed: "yes"
+        }
+      }
+      #Act
+      patch task_path(existing_task.id, params: update_task_from_data)
```

Comment:
This test doesn't use your `complete_task_path` route, and uses a completely different route/action here. In fact, it doesn't even go into the `TasksController` `toggle_complete` action here!